File: September 2022/Day 15/C++ Code/Smallest Subset with Greater Sum.cpp
//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++

class Solution{
    public:
    int minSubset(vector<int> &Arr,int N){
        sort(Arr.begin(),Arr.end());
        long long int total_sum =0;
        for(long long int n : Arr) {
            total_sum+=n;
        }
        
        long long int included_sum =0;
        for(int i=N-1;i>=0;i--) {
            included_sum+=(long long int)(Arr[i]);
            long long int excluded_sum = total_sum-included_sum;
            if(included_sum>excluded_sum) {
                return N-i;
            }
        }
        return N;
    }
};

//{ Driver Code Starts.
int main(){
    int t;
    cin>>t;
    while(t--){
        int N;
        cin>>N;
        vector<int> Arr(N);
        for(int i=0;i<N;i++){
            cin>>Arr[i];
        }
        Solution ob;
        cout<<ob.minSubset(Arr,N)<<endl;
    }
    return 0;
}
// } Driver Code Ends

File: September 2022/Day 15/JAVA Code/Smallest Subset with Greater Sum.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;
import java.util.Collections;


class GFG{
    public static void main(String args[]) throws IOException { 
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t > 0){
            int N = sc.nextInt();
            int[] Arr = new int[N];
            for (int i = 0; i < N; ++i){
                Arr[i] = sc.nextInt();
            }
            Solution ob = new Solution();
            int ans = ob.minSubset(Arr,N);
            System.out.println(ans);
            t--;
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution { 
    int minSubset(int[] Arr,int N) { 
        Arrays.sort(Arr);
        long total_sum =0;
        for(long n : Arr) {
            total_sum+=n;
        }
        
        long included_sum =0;
        for(int i=N-1;i>=0;i--) {
            included_sum+=(long)(Arr[i]);
            long excluded_sum = total_sum-included_sum;
            if(included_sum>excluded_sum) {
                return N-i;
            }
        }
        return N;
    }
}
